.specs/unsynced-covalues-tracking/design.md

```diff
@@ -0,0 +1,239 @@
+
+  this.storeContent(content);
+
+  const contentKnownState = knownStateFromContent(content);
+
+  for (const peer of this.getPeers(coValue.id)) {
+    this.unsyncedTracker.add(coValue.id, peer.id);
+
+    // ...
+
+    this.trySendToPeer(peer, content);
+    const unsubscribe = this.syncState.subscribeToPeerUpdates(
+      peer.id,
+      (knownState, syncState) => {
+        if (syncState.uploaded && knownState.id === coValue.id) {
+          this.unsyncedTracker.remove(coValue.id, peer.id);
+          unsubscribe();
+        }
+      },
+    );
+
+    peer.combineOptimisticWith(coValue.id, contentKnownState);
+    peer.trackToldKnownState(coValue.id);
+  }
+}
+ ```
+- Add method `async resumeUnsyncedCoValues()` to load and resume syncing unsynced CoValues. This happens asynchronously and doesn't block initialization
+- Call `resumeUnsyncedCoValues` as part of `SyncManager.startPeerReconciliation`.
+
+### 4. Sync Status Subscriptions
+
+**Location:** `packages/cojson/src/coValueCore/coValueCore.ts` and `packages/cojson/src/sync.ts`
+
+**CoValueCore.subscribeToSyncStatus:**
+- Subscribe to changes in whether this specific CoValue is synced
+- Uses `syncManager.unsyncedTracker.subscribe(coValueId)` to get notified when the CoValue's sync status changes
+- Calls listener immediately with current state on subscription
+
+**SyncManager.subscribeToSyncStatus:**
+- Subscribe to changes in whether all CoValues are synced
+- Uses `syncManager.unsyncedTracker.subscribe()` to get notified when the unsynced set changes
+- Calls listener immediately with current state on subscription (check `unsyncedTracker.isAllSynced()`)
+
+### 5. Refactored waitForSync
+
+**Location:** `packages/cojson/src/coValueCore/coValueCore.ts` and `packages/cojson/src/sync.ts`
+
+**Changes:**
+- Replace the current implementation of `CoValueCore.waitForSync()` based on CoValue subscription with a simple call to `syncManager.unsyncedTracker.subscribe(coValueId)`.
+
+## Data Model
+
+### UnsyncedCoValuesTracker In-Memory Representation
+
+The tracker maintains an in-memory data structure that maps each unsynced CoValue to the set of peers it's unsynced to:
+
+```typescript
+class UnsyncedCoValuesTracker {
+  // Map from CoValue ID to Set of Peer IDs that the CoValue is unsynced to
+  private unsynced: Map<RawCoID, Set<PeerID>> = new Map();
+  private coValueListeners: Map<RawCoID, (synced: boolean) => void> = new Map();
+  private globalListeners: (synced: boolean) => void = new Set();
+  private storage?: StorageAPI;
+}
+```
+
+### Storage Layout
+
+Storage persists unsynced CoValue-to-peer relationships as individual rows, with one row per (CoValue ID, Peer ID) pair.
+
+**IndexedDB Schema:**
+- Object store: `"unsyncedCoValues"`
+- Key: `[coValueId, peerId]` (composite key)
+- Value: `{ coValueId: RawCoID, peerId: PeerID }`
+- Indexes:
+  - Index on `coValueId` for efficient queries by CoValue
+  - Index on `peerId` for efficient queries by peer (optional, for cleanup)
```

Comment:
How are we going to use this `PeerID` information?

What if we do change it, let's say the dev changes the API Key?